include/TextStreamReader.h

```diff
@@ -0,0 +1,104 @@
+    /// @brief An input Stream (in text mode) helper class for safer streaming.
+    ///////////////////////////////////////
+    class MEZZ_LIB TextStreamReader
+    {
+    protected:
+        /// @brief A pointer to the Stream being read from.
+        StdInputStreamPtr Stream;
+    public:
+        /// @brief Stream constructor.
+        /// @param Input The Stream to read from.
+        TextStreamReader(StdInputStreamPtr Input);
+        /// @brief Class destructor.
+        ~TextStreamReader() = default;
+
+        ///////////////////////////////////////////////////////////////////////////////
+        // Reading
+
+        /// @brief Reads N bytes from the Stream.
+        /// @remarks This will also advance the Stream position by the number of bytes read. The read operation
+        /// may not read the amount of bytes specified. To verify, query the size on the returned String.
+        /// @param Bytes The number of bytes to read from the Stream.
+        /// @return Returns a String containing the data read from the Stream.
+        [[nodiscard]] String Read(const StreamSize Bytes);
+        /// @brief Reads characters from the Stream until the delimiter is found.
+        /// @remarks This will also advance the Stream position by the number of bytes read, which includes the
+        /// the specified delimiter. So any further read operation will begin at the Stream position just after the
+        /// delimiter. However the delimiter will not be returned as a part of the return String.
+        /// @param Delim The character to stop at when reading the Stream.
+        /// @return Returns a String with all characters from the read position to (but not including) the delimiter.
+        [[nodiscard]] String ReadLine(const Char8 Delim = '\n');
+        /// @brief Reads a single character from the Stream.
+        /// @remarks This will also advance the Stream position by 1 byte.
+        /// @return Returns the character at the current read position, or a null terminator if the Stream is at EoF.
+        [[nodiscard]] Char8 ReadCharacter();
+        /// @brief Advances the read position in the Stream.
+        /// @remarks The return and Bytes will match if the operation was successful.
+        /// @param Bytes The number of bytes to advance in the Stream.
+        /// @return The number of bytes actually skipped in the Stream.
+        StreamSize Skip(const StreamSize Bytes);
+
+        /// @brief Reads the entire Stream and places it into a String.
+        /// @remarks This function reads the entire Stream from start to finish. The read position will be saved,
+        /// then set to the beginning of the Stream, advanced to the end, and then set back to the same read
+        /// position the Stream was at when this function was called.
+        /// @n @n
+        /// This might produce a very large String. Be mindful of the Streams size when using this.
```

Comment:
You suggest being mindful of the stream size here. Yet there is no function for gathering that information. In other places you noted when you expected the game dev to know things. Should the game dev come with a good idea. should there be a size function on this, or would it be prudent to @ref some function on the InputStream?

Also, per our discussion some streams are impossible to check the size of. Is it possible to make a function to check if the stream is can be reasonably read at once?

Reply:
The check already exists.  In the third line of the function where we check the value of the returned result of tellg, we're seeing if we got a gibberish value.  If we simply cannot seek in the stream, we don't bother and immediately return an empty String.

Alternatively, the developer can check the stream before creating a Reader with it.  If it's just a standard stream then they can try to seek with it themselves, or they may also have an idea of where it came from because presumably they initialized it.  If it's one of our streams then they can query the size of the stream directly with our API.